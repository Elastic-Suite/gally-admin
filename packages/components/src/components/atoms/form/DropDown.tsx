import React, { ForwardedRef, useCallback } from 'react'

import { IFieldErrorProps, IValidator, useFormError } from '../../../hooks'

import Dropdown, { IDropDownProps } from './DropDownWithoutError'

interface IDropDownErrorProps<T> extends IFieldErrorProps, IDropDownProps<T> {}

function DropdownError<T>(props: IDropDownErrorProps<T>): JSX.Element {
  const {
    onChange,
    showError,
    additionalValidator,
    replacementErrorsMessages,
    ...inputProps
  } = props

  const validator = useCallback<IValidator>(
    (value, event) => {
      if (additionalValidator) return additionalValidator(value, event)
      if (
        inputProps.required &&
        (value === undefined ||
          value === '' ||
          (value as unknown[]).length === 0)
      ) {
        return 'valueMissing'
      }
      return ''
    },
    [additionalValidator, inputProps.required]
  )

  const [{ ref, ...formErrorProps }] = useFormError(
    onChange,
    inputProps.value,
    showError,
    validator,
    inputProps.disabled,
    replacementErrorsMessages
  )

  return (
    <Dropdown
      {...inputProps}
      {...formErrorProps}
      error={inputProps?.error || formErrorProps?.error}
      helperIcon={inputProps?.helperIcon || formErrorProps?.helperIcon}
      helperText={inputProps?.helperText || formErrorProps?.helperText}
      ref={ref as ForwardedRef<HTMLInputElement>}
    />
  )
}

export default DropdownError
