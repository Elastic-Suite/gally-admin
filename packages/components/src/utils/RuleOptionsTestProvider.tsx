import React, { ReactNode } from 'react'
import {
  IRuleEngineOperators,
  RuleAttributeType,
  getIri,
} from '@elastic-suite/gally-admin-shared'

import ruleEngineOperator from '../../public/mocks/rule_engine_operators.json'
import sourceFields from '../../public/mocks/source_fields.json'
import sourceFieldLabels from '../../public/mocks/source_field_labels.json'

import RuleOptionsProvider from '../components/stateful-providers/RuleOptionsProvider/RuleOptionsProvider'

interface IProps {
  children: ReactNode
}

const sourceFieldLabelsMap = new Map(
  sourceFieldLabels['hydra:member'].map((label) => [
    label.sourceField,
    label.label,
  ])
)

const fields = sourceFields['hydra:member'].map((field) => {
  const label = sourceFieldLabelsMap.get(getIri('source_fields', field.id))
  return {
    id: field.id,
    code: field.code,
    label: (label || field.defaultLabel) ?? '',
    type: field.type as RuleAttributeType,
  }
})

function RuleOptionsTestProvider(props: IProps): JSX.Element {
  const { children } = props
  return (
    <RuleOptionsProvider
      catalogId={-1}
      localizedCatalogId={-1}
      fields={fields}
      ruleOperators={ruleEngineOperator as IRuleEngineOperators}
    >
      {children}
    </RuleOptionsProvider>
  )
}

export default RuleOptionsTestProvider
