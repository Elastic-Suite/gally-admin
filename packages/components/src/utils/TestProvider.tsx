import React, { ReactNode } from 'react'

import catalogs from '../../public/mocks/catalog.json'
import metadata from '../../public/mocks/metadata.json'

import { catalogContext, metadataContext } from '../contexts'
import { useCatalogs } from '../hooks'

import OptionsProvider from '../components/stateful-providers/OptionsProvider/OptionsProvider'

interface IProps {
  children: ReactNode
}

function TestProvider(props: IProps): JSX.Element {
  const { children } = props
  const contextValueCatalog = useCatalogs(catalogs['hydra:member'])
  const contextValueMetadata = metadata['hydra:member']

  return (
    <OptionsProvider>
      <catalogContext.Provider value={contextValueCatalog}>
        <metadataContext.Provider value={contextValueMetadata}>
          {children}
        </metadataContext.Provider>
      </catalogContext.Provider>
    </OptionsProvider>
  )
}

export default TestProvider
