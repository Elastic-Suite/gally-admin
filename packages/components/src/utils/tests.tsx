import React, { ReactNode } from 'react'
import { RenderOptions, render, renderHook } from '@testing-library/react'
import { PreloadedState } from '@reduxjs/toolkit'
import { Bundle, api } from '@elastic-suite/gally-admin-shared'

import { AppStore, RootState, setupStore } from '../store'

import AppProvider from '../components/stateful-providers/AppProvider/AppProvider'

import TestProvider from './TestProvider'

export interface IExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
}

export interface IRenderOutput {
  store: AppStore
}

const defaultState = {
  data: {
    api,
    bundles: [Bundle.VIRTUAL_CATEGORY],
    configurations: {
      'base_url/media': 'https://localhost/media/catalog/product/',
    },
    metadata: [
      {
        '@id': '/metadata/3',
        '@type': 'Metadata',
        id: 3,
        entity: 'product',
        sourceFields: ['/source_fields/217', '/source_fields/253'],
      },
      {
        '@id': '/metadata/4',
        '@type': 'Metadata',
        id: 4,
        entity: 'category',
        sourceFields: ['/source_fields/254', '/source_fields/286'],
      },
    ],
  },
  user: {
    requestedPath: '',
  },
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = defaultState,
    store,
    ...renderOptions
  }: IExtendedRenderOptions = {}
) {
  store = store ?? setupStore(preloadedState)
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return (
      <AppProvider store={store}>
        <TestProvider>{children}</TestProvider>
      </AppProvider>
    )
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderHookWithProviders<R>(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  hook: (props: any) => R,
  {
    preloadedState = defaultState,
    store,
    ...renderOptions
  }: IExtendedRenderOptions = {}
) {
  store = store ?? setupStore(preloadedState)
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return (
      <AppProvider store={store}>
        <TestProvider>{children}</TestProvider>
      </AppProvider>
    )
  }
  return { store, ...renderHook(hook, { wrapper: Wrapper, ...renderOptions }) }
}
