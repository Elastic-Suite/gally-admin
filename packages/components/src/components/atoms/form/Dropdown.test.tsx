import React from 'react'

import { renderWithProviders } from '../../../utils/tests'

import DropDown from './DropDown'

describe('DropDown match snapshot', () => {
  it('BadgeDisabledFalse', () => {
    const { container } = renderWithProviders(
      <DropDown
        disabled={false}
        options={[
          { label: 'Ten', value: 10 },
          { label: 'Twenty', value: 20 },
          { label: 'Thirty', value: 30 },
          { label: 'Forty', value: 40, disabled: true },
        ]}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('BadgeDisabledTrue', () => {
    const { container } = renderWithProviders(
      <DropDown
        disabled
        options={[
          { label: 'Ten', value: 10 },
          { label: 'Twenty', value: 20 },
          { label: 'Thirty', value: 30 },
          { label: 'Forty', value: 40 },
        ]}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('BadgeLabel', () => {
    const { container } = renderWithProviders(
      <DropDown
        label="Label"
        options={[
          { label: 'Ten', value: 10 },
          { label: 'Twenty', value: 20 },
          { label: 'Thirty', value: 30 },
          { label: 'Forty', value: 40 },
        ]}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('BadgeRequiredFalse', () => {
    const { container } = renderWithProviders(
      <DropDown
        required={false}
        options={[
          { label: 'Ten', value: 10 },
          { label: 'Twenty', value: 20 },
          { label: 'Thirty', value: 30 },
          { label: 'Forty', value: 40 },
        ]}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('BadgeRequiredTrue', () => {
    const { container } = renderWithProviders(
      <DropDown
        required
        options={[
          { label: 'Ten', value: 10 },
          { label: 'Twenty', value: 20 },
          { label: 'Thirty', value: 30 },
          { label: 'Forty', value: 40 },
        ]}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('BadgeMultiple', () => {
    const { container } = renderWithProviders(
      <DropDown
        value={[]}
        multiple
        disabled={false}
        label="Label"
        options={[
          { label: 'Ten', value: 10 },
          { label: 'Twenty', value: 20 },
          { label: 'Thirty', value: 30 },
          { label: 'Forty', value: 40 },
        ]}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('BadgeArrayObjectValue', () => {
    const { container } = renderWithProviders(
      <DropDown
        value={[
          { '@id': 1, locale: 'fr_FR' },
          { '@id': 2, locale: 'en_US' },
        ]}
        multiple
        disabled={false}
        label="Label"
        objectKeyValue="locale"
        options={[
          { value: 'en_US', label: 'English (United States)' },
          { value: 'fr_FR', label: 'French (France)' },
          { value: 'en_CA', label: 'English (Canada)' },
        ]}
      />
    )
    expect(container).toMatchSnapshot()
  })
  it('BadgeObjectValue', () => {
    const { container } = renderWithProviders(
      <DropDown
        value={{ '@id': 1, locale: 'fr_FR' }}
        disabled={false}
        label="Label"
        objectKeyValue="locale"
        options={[
          { value: 'en_US', label: 'English (United States)' },
          { value: 'fr_FR', label: 'French (France)' },
          { value: 'en_CA', label: 'English (Canada)' },
        ]}
      />
    )
    expect(container).toMatchSnapshot()
  })
})
