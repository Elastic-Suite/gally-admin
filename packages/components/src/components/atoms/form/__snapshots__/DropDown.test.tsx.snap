// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDown sould match snapshot 1`] = `
<div>
  <div
    class="MuiFormControl-root css-1girsym-MuiFormControl-root"
  >
    <div
      class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon css-1qqsdnr-MuiAutocomplete-root"
    >
      <div
        class="MuiFormControl-root css-hh3vi8-MuiFormControl-root"
      >
        <div
          class="InputText__Wrapper css-17fp4jk"
        >
          <div
            class="MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-33c5sc-MuiInputBase-root"
          >
            <input
              aria-autocomplete="list"
              aria-expanded="false"
              aria-invalid="false"
              autocapitalize="none"
              autocomplete="off"
              class="MuiInputBase-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-ngazp5-MuiInputBase-input"
              id=":r0:"
              role="combobox"
              spellcheck="false"
              type="text"
              value="Test 1"
            />
            <div
              class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
            >
              <button
                aria-label="form.clear"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-clearIndicator css-1glvl0p-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-clearIndicator"
                tabindex="-1"
                title="form.clear"
                type="button"
              >
                <ion-icon
                  name="close"
                />
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="form.open"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                tabindex="-1"
                title="form.open"
                type="button"
              >
                <ion-icon
                  name="chevron-down"
                />
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DropDown sould match snapshot width multiple value 1`] = `
<div>
  <div
    class="MuiFormControl-root css-1girsym-MuiFormControl-root"
  >
    <div
      class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon css-1qqsdnr-MuiAutocomplete-root"
    >
      <div
        class="MuiFormControl-root css-hh3vi8-MuiFormControl-root"
      >
        <div
          class="InputText__Wrapper css-17fp4jk"
        >
          <div
            class="MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-33c5sc-MuiInputBase-root"
          >
            <div
              class="MuiButtonBase-root MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-deletable MuiChip-deletableColorDefault MuiChip-filledDefault MuiAutocomplete-tag MuiAutocomplete-tagSizeMedium css-69f6vo-MuiButtonBase-root-MuiChip-root"
              data-tag-index="0"
              role="button"
              style="opacity: 1;"
              tabindex="-1"
            >
              <span
                class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
              >
                Test 1
              </span>
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiChip-deleteIcon MuiChip-deleteIconMedium MuiChip-deleteIconColorDefault MuiChip-deleteIconFilledColorDefault css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <ion-icon
                  name="close"
                />
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <div
              class="MuiButtonBase-root MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-deletable MuiChip-deletableColorDefault MuiChip-filledDefault MuiAutocomplete-tag MuiAutocomplete-tagSizeMedium css-69f6vo-MuiButtonBase-root-MuiChip-root"
              data-tag-index="1"
              role="button"
              style="opacity: 1;"
              tabindex="-1"
            >
              <span
                class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
              >
                Test 2
              </span>
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiChip-deleteIcon MuiChip-deleteIconMedium MuiChip-deleteIconColorDefault MuiChip-deleteIconFilledColorDefault css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <ion-icon
                  name="close"
                />
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <input
              aria-autocomplete="list"
              aria-expanded="false"
              aria-invalid="false"
              autocapitalize="none"
              autocomplete="off"
              class="MuiInputBase-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-ngazp5-MuiInputBase-input"
              id=":r1:"
              role="combobox"
              spellcheck="false"
              type="text"
              value=""
            />
            <div
              class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
            >
              <button
                aria-label="form.clear"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-clearIndicator css-1glvl0p-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-clearIndicator"
                tabindex="-1"
                title="form.clear"
                type="button"
              >
                <ion-icon
                  name="close"
                />
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="form.open"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                tabindex="-1"
                title="form.open"
                type="button"
              >
                <ion-icon
                  name="chevron-down"
                />
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
